src/common.rkt

```diff
@@ -104,6 +100,72 @@
   (check-true (set-disjoint? '() '()))
   (check-false (set-disjoint? '(a b c) '(a))))
 
+(define (subsequence? v l)
+  (or
+   (empty? v)
+   (let ([v* (member (first v) l)])
+     (and v* (subsequence? (rest v) v*)))))
+
+(module+ test
+  (define l (range 10))
+  (check-true (subsequence? empty empty))
+  (check-true (subsequence? empty l))
+  (check-true (subsequence? '(1) l))
+  (check-true (subsequence? '(1 2) l))
+  (check-true (subsequence? '(1 3 5 7 9) l))
+  (check-true (subsequence? '(1 2 5 8) l))
+  (check-false (subsequence? '(x y) l))
+  (check-false (subsequence? '(1 2 10) l)))
+
+(define (list-ref* l p)
+  (let loop ([l l] [p p] [i 0])
+    (if (or (empty? p) (empty? l))
+        empty
+        (let* ([j (first p)]
+               [k (- j i)]
+               [l* (drop l k)])
+          (cons (first l*) (loop (rest l*) (rest p) (+ i k 1)))))))
```

Comment:
1. Let's use `match` instead of `if`/`let`
2. If `l` is empty but `p` isn't, isn't that an error? Let's just make it a `match` failure
3. All this logic to use `drop` and tweak the indices all comes from the fact that you want to output a `cons`. It makes it hugely complex. Do this:

```
(reap [sow]
  (let loop ([l l] [p p] [i 0])
    (match p
      ['() (void)]
      [(cons (= i) rest)
       (sow (car l))
       (loop (cdr l) rest (+ i 1))]
      [(cons j rest)
       (loop (cdr l) p (+ i 1))])
```

I didn't test this code so uhh make sure it works.

Reply:
But also—like, this isn't performance sensitive and `(map (curry list-ref l) p)` also seems good? Maybe just inline that into the one place this is used to make sure it won't accidentally be used in a performance-sensitive place?